### ISSUE TITLE
[Question] Any thoughts using templating values for subcharts?

### ISSUE THREAD

**Issue by dtshepherd** [roboll/helmfile#355]
So this may be a bit crazy, but I'm wondering if there is a good pattern for templated subcharts...  Say I have a **chart A** that uses `helmfile` and has a `values.yaml.gotmpl` file.  Its directory structure might look like this:
```
A/
  templates/...
  Chart.yaml
  helmfile.yaml
  values.yaml
  values.yaml.gotmpl
```
This chart can exist on its own, so I want it to have its own helmfile and templated values.yaml input.

Then, I may want to reuse this chart in some other parent **chart B** (pulled in using requirements.yaml):
```
B/
  templates/...
  charts/
    B.tgz
  Chart.yaml
  helmfile.yaml
  requirements.yaml
  values.yaml
  values.yaml.gotmpl
```
The `helm package` command will package any files in the chart directory into the `tgz`, so the subchart B `values.yaml.gotmpl` will be packed with the chart.

Currently, when installing **chart B**, the `values.yaml.gotmpl` in **chart A** will be ignored.  It may contain some critical environment values needed to function properly.  Right now, we'll need to duplicate those values into **chart B**'s `values.yaml.gotmpl` like so:
````
foo: {{ .Environment.Values.foo }}

A:
  bar: {{ .Environment.Values.bar }}
```
Most likely the `A:` subsection of the `values.yaml.gotmpl` in **chart B** will be a near duplicate of the `values.yaml.gotmpl` in **chart A**.  It would be nice if we could recursively unpack the subcharts and make use of those values.yaml files when the parent chart is installed.

Again, this may be crazy, but I'd really like to DRY up some of these templates if possible.

**Comment by dtshepherd** [comment]
I’m also aware that I could just opt out of using subcharts and implement that same idea using a helmfile.  However, it’s nice to use built in helm features and helm also supports packaging and repos of the charts/subcharts together.

**Comment by mumoshu** [comment]
@dtshepherd Hey! Thanks for sharing an interesting use-case. I'm still trying to understand, but anyway - you meant `A.tgz` under the `charts/` directory in your example, rather than `B.tgz`, rght?

**Comment by dtshepherd** [comment]
Whoops, yes, just updated the comment.

**Comment by dtshepherd** [comment]
Ultimately, I think I can accomplish the same thing as subcharts with the `helmfile:` directive and some extra tooling in our build system to build an umbrella helmfile, but that doesn’t seem like a great solution.  

I started prototyping with helmfile because helm decided they were never going to allow templates values.yaml files:  https://github.com/helm/helm/issues/2492

**Comment by dtshepherd** [comment]
😂 I’m guessing that’s why 90% of us our here.

I was thinking a first pass of this feature could recursively unpack the subcharts and include their helmfile.yaml in the parent.  However, the rendered subchart values would need to be nested in the generated values.yaml under a subchart named key.

**Comment by mumoshu** [comment]
> Ultimately, I think I can accomplish the same thing as subcharts with the helmfile: directive and some extra tooling in our build system to build an umbrella helmfile, but that doesn’t seem like a great solution.

Would you mind sharing why you thought it isn't a great solution?

I still think `helmfiles:` is what you're looking for. You'd probably write `B/helmfile.yaml` that looks like the below to achieve your use-case:

```yaml
helmfiles:
- ../A/helmfile.yaml

releases:
- name: B
   chart: ./
```

**Comment by dtshepherd** [comment]
Yes, your example is probably what it will look like and that's my plan right now.  However, the reason I don't like that solution is the lack of chart versioning.  If we make use of subcharts, we can use the `requirements.yaml` to define the subchart names/versions.  In your example, I'd need to have our build system pull those dependencies using some outside tooling.

**Comment by mumoshu** [comment]
> I'd need to have our build system pull those dependencies using some outside tooling.

Ah, so you have a hard requirement that `A` and `B` must NOT reside within the same git repo, right?

**Comment by dtshepherd** [comment]
Yes, I think #330 helps here.

Ultimately, I don't need most of what helmfile provides and I was just trying to shoehorn it into supporting templating values.yaml so I can encapsulate subcharts properly.  Recently, I decided to start working on a prototype helm plugin that supports rendering templated values.yaml in a given chart (and its subcharts).  Its based on the old helm-template plugin (https://github.com/technosophos/helm-template) and uses https://github.com/helm/helm/tree/master/pkg/chartutil to process the charts values.yaml and subcharts.

At this point, you can probably close this issue if you don't have much interest in supporting it.

**Comment by mumoshu** [comment]
Nice. The simpler the better 😄 

I'm still interested in this issue. It is just that I'm not yet sure what would be the feature that fully supports this looks like.

I think your plugin would also allow me to see what it might look like when done in helmfile. Please don't hesitate to share it!

**Comment by dtshepherd** [comment]
Will do!  Need to get client approval to open source :(
